### ISSUE TITLE
Invalid recurrence pattern (?) crashes Thunderbird

### ISSUE THREAD

**Issue by jobisoft** [jobisoft/TbSync#41]
If an EAS item has the following recurrence pattern:

```
<Recurrence xmlns='Calendar'>
<Type xmlns='Calendar'>1</Type>
<Interval xmlns='Calendar'>1</Interval>
<Until xmlns='Calendar'>20110109T003000Z</Until>
<DayOfWeek xmlns='Calendar'>127</DayOfWeek>
</Recurrence>
```

Thunderbird throws

```
NS_ERROR_OUT_OF_MEMORY: Component returned failure code: 0x8007000e (NS_ERROR_OUT_OF_MEMORY) [calIRecurrenceItem.getOccurrences]
```

and does not display any events anymore. To test this, I injected the following event:

```
<Add>
<ServerId>43:1201</ServerId>
<ApplicationData>
<TimeZone xmlns='Calendar'>LAEAAEUAYQBzAHQAZQByAG4AIABTAHQAYQBuAGQAYQByAGQAIABUAGkAbQBlAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAsAAAABAAIAAAAAAAAAAAAAACgAVQBUAEMALQAwADUAOgAwADAAKQAgAEUAYQBzAHQAZQByAG4AIABUAGkAbQBlACAAKABVAFMAIAAmACAAQwAAAAMAAAACAAIAAAAAAAAAxP///w==</TimeZone>
<DtStamp xmlns='Calendar'>20120529T035130Z</DtStamp>
<StartTime xmlns='Calendar'>20101113T003000Z</StartTime>
<Subject xmlns='Calendar'>Subject</Subject>
<UID xmlns='Calendar'>nada@google.com</UID>
<OrganizerName xmlns='Calendar'>Name</OrganizerName>
<OrganizerEmail xmlns='Calendar'>Name@provider.net</OrganizerEmail>
<Location xmlns='Calendar'/>
<EndTime xmlns='Calendar'>20101113T004500Z</EndTime>
<Recurrence xmlns='Calendar'>
<Type xmlns='Calendar'>1</Type>
<Interval xmlns='Calendar'>1</Interval>
<Until xmlns='Calendar'>20110109T003000Z</Until>
<DayOfWeek xmlns='Calendar'>127</DayOfWeek>
</Recurrence>
<Body xmlns='AirSyncBase'>
<Type xmlns='AirSyncBase'>1</Type>
<EstimatedDataSize xmlns='AirSyncBase'>4</EstimatedDataSize>
<Data xmlns='AirSyncBase'>

</Data>
</Body>
<Sensitivity xmlns='Calendar'>2</Sensitivity>
<BusyStatus xmlns='Calendar'>2</BusyStatus>
<AllDayEvent xmlns='Calendar'>0</AllDayEvent>
<MeetingStatus xmlns='Calendar'>0</MeetingStatus>
<NativeBodyType xmlns='AirSyncBase'>3</NativeBodyType>
<DisallowNewTimeProposal xmlns='Calendar'>0</DisallowNewTimeProposal>
<ResponseRequested xmlns='Calendar'>1</ResponseRequested>
<ResponseType xmlns='Calendar'>0</ResponseType>
</ApplicationData>
</Add>
```

Is this a bug in Thunderbird? Can Thunderbird create such a pattern by itself? What to do?

**Comment by ChrisJAllan** [comment]
DayOfWeek = 127 is a special value for monthly and yearly occurrences for the last day of the month. Is it supposed to be treated literally for weekly? How do other systems handle such an event?

**Comment by jobisoft** [comment]
This is a live event coming from a calendar of one of my users, I will check what server it is. It looks like the type is wrong, because "end of week" would be simply "every Sunday" which would look different.

I will check, what the calender of the user is actually saying.

**Comment by ChrisJAllan** [comment]
Tested against Outlook, it is treated literally. PR incoming.

**Comment by jobisoft** [comment]
The user reported, the original pattern is:

Périodicity is "A lieu chaque dimanche, lundi, mardi, mercredi, jeudi, vendredi et samedi à compter du 2010-11-12 jusqu'au 2011-01-08 de 19:30 à 19:45."
Which is in English "It takes place every Sunday, Monday, Tuesday, Wednesday, Thursday, Friday and Saturday from 2010-11-12 until 2011-01-08 from 19:30 to 19:45."

Does that match your findings?



Am 30.01.18 um 00:31 schrieb Chris Allan

**Comment by ChrisJAllan** [comment]
Yup, a 'weekly' event that happens on every day.

**Comment by jobisoft** [comment]
Ah, now I get it, 127 has the special meaning ONLY when monthly or yearly, otherwise each bit of 127 is interpreted (which is everyday), right?

Is your recurrence generator also able to create such an event correctly?



Am 30.01.18 um 10:04 schrieb Chris Allan

**Comment by ChrisJAllan** [comment]
Correct. And yes, it is sent correctly. (I was testing this by first creating the event in TB, syncing it, making a meaningless modification on the server, then syncing it back.)

**Comment by jobisoft** [comment]
probably fixed by #42
